hordak/models/core.py

```diff
@@ -579,7 +599,10 @@ class Leg(models.Model):
     def save(self, *args, **kwargs):
         if self.amount.amount == 0:
             raise exceptions.ZeroAmountError()
-        return super(Leg, self).save(*args, **kwargs)
+
+        leg = super(Leg, self).save(*args, **kwargs)
+        transaction.on_commit(lambda: _enforce_leg(transaction_id=self.transaction_id))
```

Comment:
Maybe I missed something, but is this check not executed after the transaction is committed?
From https://docs.djangoproject.com/en/4.2/topics/db/transactions/#performing-actions-after-commit
> [on_commit()](https://docs.djangoproject.com/en/4.2/topics/db/transactions/#django.db.transaction.on_commit) allows you to register callbacks that will be executed after the open transaction is successfully committed

Reply:
hmm that is possible actually, it's a bit tricky though as if `_enforce_leg` is called before the transaction is committed, changes made won't be visible to the `check_leg` stored procedure. I wonder if a suitable solution would be to rollback the transaction if the procedure fails?

Reply:
Afaik you see the changes "in" a transaction even before it is committed. 
> I wonder if a suitable solution would be to rollback the transaction if the procedure fails?

I don't think that you can rollback a already committed transaction.

Reply:
> Afaik you see the changes "in" a transaction even before it is committed.
> 
> > I wonder if a suitable solution would be to rollback the transaction if the procedure fails?
> 
> I don't think that you can rollback a already committed transaction.

I will have a look into this. In the meantime would you mind opening this as a new issue? Commenting on an already-closed PR might mean this issue has poor visibility.

Reply:
After looking at this again, I think the reason I did this in `on_commit` was to emulate the behaviour of the original Postgres trigger:
https://github.com/adamcharnock/django-hordak/blob/master/hordak/migrations/0002_check_leg_trigger_20160903_1149.py

Postgres has [deferred triggers](https://www.postgresql.org/docs/9.1/sql-createtrigger.html), which Hordak uses to check the legs are equal to zero *after* a transaction has taken place (need to make sure both legs have been added first), but MySQL doesn't have this.

Reply:
Apologies for being pedantic. I believe it's still inside the transaction:

> They can be fired either at the end of the statement causing the triggering event, or at the end of the containing transaction; in the latter case they are said to be deferred.

Reply:
It's inside the transaction, but after all the changes have been applied, which isn't something MySQL has, so we'd have to emulate this behaviour somehow

Reply:
I've just added a PR for a possible approach to improving on this behaviour:
#106 

Currently it requires you to decorate the core functions with `@enforce_atomic` (replacing `@db_transaction.atomic`)
Some of the tests will fail currently as they don't use this decorator. Ideally there would be a way of enforcing this without the decorator.